### ISSUE TITLE
Experiment with unicode mode in the Editor

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1527]
When I was writing #1297 I was not at all sure about how many tasks and what effort overall it would take to move AGS to unicode.
But a success at featuring unicode mode in the engine (#1321) have resolved this problem by 1/2 (or maybe even 2/3).

After #1321 engine can now work in ASCII or UTF-8 text mode, and switch that mode freely at runtime. Currently this is only done when loading a translation which has a "encoding" switch, used as a signal for the engine. But there's no technical necessity to rely on translation file: that was done mostly because translations are the simpliest available data files that may be edited outside of the editor, and did not require much changes to compile from utf-8 text source.

I left this topic afterwards, being distracted by other priorities at the time, but kept thinking that *maybe* we are closer to the unicode support in the Editor than I initially assumed.

---

So I suggest doing an experiment. An important reminder though: as we still have to maintain backward compatibility here, the unicode support must be done through a switch. Presumably this switch may be located right in the General Settings, and set to "ANSI" mode for the older imported projects.

Let's briefly recap the Editor's situation from the point of text encoding. There are roughly 4 components which pass (text) data between each other:

1. Editor input controls. For *text* these are: property grid, and Scintilla text editor used for scripts. Because Editor is a .NET program it supports unicode, so property grid has to be working in unicode. I don't remember and not certain about Scintilla: it could be unicode-aware, or having an encoding setting. In the worst case it's ansi-only, which would suck, but I don't think that may be the case?...
2. "Native" data structures in the C++ library. Now, these work with pure char (byte) arrays and strings, which may contain anything, it's mostly a question of how to interpret that (see what was done for the engine).  The problem is *how the text data is converted when it's passed from .NET side* (and back). IIRC this is done using special functions that convert managed strings into native ones through the *chosen Encoding object*. Supposedly, if we replace one Encoding object with another (depending on a setting value), we may convert C# Strings into UTF-8 byte sequence nicely.
3. Data in compiled game files. Once more, if these are written by .NET side, then text data is converted using a chosen Encoding object. Replacing one should affect the encoding at destination. Those which are written by the native C++ part will keep the strings and char arrays as-is, no questions asked (and these were already encoded correctly in p2).
4. Script compiler. Here, I have no idea how things are. Assuming we made Scintilla save text in utf-8, and the text is also passed into the compiler as utf-8; utf-8 does not use 0 char for anything special other than the string terminator, ~but utf-8 characters may consist of bytes that would mean something for compiler if taken out of utf-8 context. Like, matching a code for a bracket, a double-quote symbol, and so on. This likely means that compiler *must* be unicode-aware, or rather have a switch between ascii and utf-8 modes similar to the engine~. **UPDATE:** i keep forgetting the utf-8 format; correction: multibyte utf-8 sequences cannot contain byte values < 128, which makes them impossible to conflict with ascii chars used in program syntax.

Looking at above, it seems that the script compiler is the highest priority problem. Not only it may require most changes, but also scripts usually contain most of the text data in game. If compiler won't be done, doing the rest makes little sense. Therefore I suggest checking it out first. Perhaps, compiler could reuse same method as the engine uses now, although engine uses unicode functions from allegro4 library, and compiler does not have allegro4 dependency.

Not assigning this to any milestone for now.

---

Recording necessary changes:

- [ ] In Editor's C# code has all text data written into the compiled data converted using a "converter object", which either has a switch or is a virtual interface with implementations for ASCII and Unicode mode. This way we wont have to make such switch in each place.
- [ ] Similarily, has this in AGS.Native (C++ CLR code), where a "converter object" helps convert from managed strings to C ones using again either a switch or virtual inheritance.
- [ ] In Editor ensure that all the "script name" properties are safeguarded to have only characters in the range allowed by the syntax (a-z, A-Z, 0-9, and underscore).
- [ ] Add encoding to the Scripts metadata in the game project, and use it when loading/saving script source code.

**Comment by rofl0r** [comment]
i just did a quick test with agscc (based on 3.5.1 script compiler:

```
$ cat testutf.c
void print_utfd() {
        System.Log(1, "hähähäh");
}
$ ../ascc testutf.c
$ agsdisas testutf.o
warning: 6 unusual bytes in string data section, file may be corrupted
$ cat testutf.s | head -n 20
.text

print_utfd$0: ; 0 args
        thisaddr 0
        li ax, "hähähäh"
        farpush ax
        li ax, 1
        farpush ax
        setfuncargs 2
        li ax, System::Log^102
        farcall ax
        farsubsp 2
        ret 
.strings
"hähähäh"
.imports
000000000000""
000000000001""
000000000002""
000000000003""
```

it seems there's nothing needed on the compiler side, strings are simply treated as a series of bytes and saved in their binary representation.

**Comment by ivan-mogilko** [comment]
Ah! I keep forgetting the specifics of utf-8 format. So, in utf-8 character only the first byte may have values < 127, in which case that's the only byte. And if the character is represented by a multibyte sequence, then all of these bytes must have values > 127, which makes them impossible to conflict with the ascii range of symbols.

**Comment by rofl0r** [comment]
> EDIT: actually, i'm not sure anymore if there will be any problems if these are even used in symbol names. Maybe not.

that's a quite bad idea and should be discouraged (if needed, with an added compiler check). allowing unicode symbol names can lead to all sorts of issues, starting with the lexer, and ending with 3rd party tooling. and it's totally gratuitous, because it's not visible to the game user.

**Comment by ericoporto** [comment]
@ivan-mogilko  Scintilla has to receive [SC_CP_UTF8](https://github.com/adventuregamestudio/ags/blob/638b50fa29f15af13297edd708afedc59d56db07/Editor/AGS.Controls/ScintillaNET/Constants.cs#L56) when setting the [codepage](https://github.com/adventuregamestudio/ags/blob/638b50fa29f15af13297edd708afedc59d56db07/Editor/AGS.Controls/ScintillaNET/ScintillaControl.Generated.cs#L576). Not sure how to handle this, if a property in the project general settings.

**Comment by ivan-mogilko** [comment]
So, I managed to make a crude UTF-8 test for the script editor: https://github.com/ivan-mogilko/ags-refactoring/commits/experiment--unicodeeditor

I split changes into small commits to better illustrate the existing problems.
* Switch in General Settings that saves an option into game data, letting engine know to switch to utf-8 on game load.
* Scintilla editor: besides explicit unicode switch, it also had couple of hacks in its internal code, possibly introduced by CJ; according to comments this was done to allow extended ASCII chars work.
* Scripts: must save and load data as utf-8. This is done very crudely now. I think it may be possible to add a proper option into Script meta data in the project, telling which encoding to use.
* Script conversion from C# to native memory. Again, done dirty; for easier maintenance we could perhaps feature a "test converter object" that works depending on a setting.

**Comment by fernewelten** [comment]
> that's a quite bad idea and should be discouraged (if needed, with an added compiler check)

It is. Both compilers use their own routines to check what "alphabetic" or "alphanumeric" characters are. Namely `[A-Za-z}` and `[A-Za-z0-9}`, respectively. (As concerns the "classic" compiler, this is AFAICR.) So this won't change even if the locale is changed. If you declare variables or functions such as `Bräustübl` in AGS code, you get syntax errors.

However, when we enable Unicode in the Editor, we'd have to make sure that the Properties grid of the Editor still enforces these rules when identifiers are concerned, e.g. object or view or font names.

**Comment by rofl0r** [comment]
> in identifiers then they will balk at the second byte of the umlaut and name that as the culprit.

hmm, i think the simplest solution here is to just print the entire string, e.g.:

`line 104: 'Bräustübl': invalid symbol name`

so we don't have to add proper utf-8 parsing to the compiler.

> And if the source files can start with a BOM then that must be stripped before the source is handed over to the compiler.

we've had the BOM discussion elsewhere and the consensus was that it shouldn't be used. and if scintilla doesn't emit it, it should be fine because i assume people write their code in the IDE.

**Comment by ivan-mogilko** [comment]
Updated the experimental branch:
https://github.com/ivan-mogilko/ags-refactoring/commits/experiment--unicodeeditor

Unless I missed something, it seem generally working! (still a little dirty at places, especially where editor triggers ascii/utf mode change).

Tested working:
* Texts in general settings and global objects (characters etc);
* Texts in room objects (descriptions)
* Texts in script (already tested before).

**Comment by rofl0r** [comment]
cool. i noticed there are some whitespace issues whenever `tcv.ConvertASCII` was inserted, it appears leading tabs have been converted partially to spaces (most visible in MSSSCPI.cs if i recall the filename right)

**Comment by ivan-mogilko** [comment]
For backward compatibility, I've been thinking of providing a list of ansi codepages to compile the game with. At the moment AGS Editor uses just the current system setting. This is fine (sort of) while there's 1 person working on a game, but becomes bug prone when there are multiple contributors with different system locales. Besides, it's implicit and there's no direct control over it in AGS, you have to go to system settings and change "locale for non-unicode programs".

Having such setting in the Editor _might be_ technically trivial, as C# features a list of Encoder objects so it's mostly a matter of how to assign one for the use. Suppose it will have something like "Unicode (UTF-8)" and then the list of known ANSI codepages, with the ability to choose current system's one too.

On another hand, I suspect it may also be safe to always save game project (xml) in unicode encoding, regardless of how the game is compiled. As this game data is being converted from unicode anyway (as users enter text in unicode-string fields).

**Comment by rofl0r** [comment]
> For backward compatibility, I've been thinking of providing a list of ansi codepages to compile the game with. At the moment AGS Editor uses just the current system setting. This is fine (sort of) while there's 1 person working on a game, but becomes bug prone when there are multiple contributors with different system locales.

i think the best solution would be to just save the default locale name of the user that created the initial project when ANSI type is selected, so it would behave the same everywhere.

**Comment by ivan-mogilko** [comment]
> i think the best solution would be to just save the default locale name of the user that created the initial project when ANSI type is selected, so it would behave the same everywhere.

Yes, the plan is to assign this as a default value for the new projects.

However, we must also expect users upgrading the older (ANSI) project to Unicode.

---

Right now, my biggest concern remaining is the data transfer between projects, whether that would be a full project import or importing/exporting individual files (rooms, scripts, etc). For such transfer to work correctly each of these files should have an encoding hint.

What has this "hint":
* Game project file (Game.agf), in the xml header (and will also have this new setting).
* Various game data which may be exported is also saved in xml with the encoding in header.
* Translation files may have this optionally since #1321, in both source and compiled form.

What _does not_ have this "hint":
* Room files. They are still are a binary files in the project source. It may be possible to add an extension with "string options", similar to what I added to TRA format, and this extension may contain encoding hint.
* The packed script modules. These *may* have meta data, like currently they have description and author's name, so probably it's possible to add an encoding hint.
* Raw script sources. This is the "worst" case, as script format is a text with no obligation to have any "meta data" in them. We kind of could support an optional hint, like in the comments at the top of the script, but can't expect its presence. Right now the Editor does not allow import raw script files for some silly reason (see #589), but this may become an issue in the future. I guess for now we'd have to just assume the script encoding matches the game setting.

**Comment by rofl0r** [comment]
maybe you're overthinking this ?
imo there's only 2 possible supportable scenarios

- entire project is ansi
- entire project is utf8

in both cases it's sufficient to save a single global setting.

for someone wanting to convert their old project, i think it would be reasonable to except them to just throw the switch in the IDE from ansi to utf-8 and have them fix all strings manually. trying to automatically cover all cases where there could be mismatching locale in any file would seem to result in a ton of code complexity having to be added , with tons of bugs and issues coming along them.

**Comment by rofl0r** [comment]
> I would not say it would require alot of additional code though, it's mostly a matter of selecting encoder/converter object and passing into a data reading function

hmm, i've been refering mostly to plans to introduce new elements in room files, new version numbers and the like. but also imagine e.g. a room has 20 different proerpties with a string in them, now you have to wrirte code like `room.property_x = iconv("utf8", "CP8950", room.property_x); ` for every single one of those properties.

**Comment by ivan-mogilko** [comment]
> imagine e.g. a room has 20 different proerpties with a string in them, now you have to wrirte code like room.property_x = iconv("utf8", "CP8950", room.property_x); for every single one of those properties.

In the editor it does that already; that's inevitable as the room is stored as already a runtime data in the project, so it has to convert back and forth on load/save.

In the engine, I guess, we may ignore this.

**Comment by ivan-mogilko** [comment]
After investigating the editor's code and experimenting bit more, I think it's possible to separate this into two tasks; and maybe two general PRs:

1. Implement a text format switch in the project, and make everything use the given encoding when saving text & compiling the game. This should be enough for starting a new project.
2. Implement a full project conversion from one encoding to another upon format switch. I currently believe that it's possible to do without adding much code, as, like mentioned above, editor processes game text anyway, so that would be a matter of replacing hardcoded encoding with the variable one. If i'm right, then the conversion could be implemented "simply" as loading each source file that may contain text into memory and then saving it. Loading will be performed using old encoding, and saving using new encoding. The files that has to be reencoded are: Game.agf (xml), scripts and room files. Translation sources (TRS) likely should not be reencoded automatically, because they may depend on individual settings and fonts; so i'd leave them for users to convert if they need to.

**Comment by ivan-mogilko** [comment]
Resolved by #1542, #1544 and #1548.
